File: DailyChallenges/Scala/LeetCode_380.scala
import scala.collection.mutable.{Map, ArrayBuffer}
import scala.util.Random

class RandomizedSet() {
    private val m = Map.empty[Int, Int]
    private val a = ArrayBuffer.empty[Int]

    def insert(`val`: Int): Boolean = m.get(`val`) match {
        case Some(_) => false
        case None => 
            a.append(`val`)
            m += `val` -> (a.length-1)
            true
    }

    def remove(`val`: Int): Boolean = m.get(`val`) match {
        case Some(idx) =>
            a(idx) = a.last
            m(a.last) = m(`val`)
            m.remove(`val`)
            a.remove(a.length-1)
            true
        case None => false
    }

    def getRandom(): Int = {
        a(Random.nextInt(a.length))
    }
}

File: DailyChallenges/CPP/LeetCode_380.cpp
// 380. Insert Delete GetRandom O(1) @ 2022/11/29

class RandomizedSet {
private:
    vector<int> a;
    unordered_map<int, int> m;
    
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (m.find(val) != m.end())
            return false;
        else{
            a.push_back(val);
            m[val] = a.size()-1;
            return true;
        }
    }
    
    bool remove(int val) {
        if (m.find(val) == m.end())
            return false;
        else{
            m[a.back()] = m[val];
            a[m[val]] = a.back();
            m.erase(val);
            a.pop_back();
            return true;
        }
    }
    
    int getRandom() {
        return a[rand()%a.size()];
    }
};
